Add `start_day` parameter to `Time#at_beginning_of_week`

DanielGilchrist opened the pull request:
Closes https://github.com/crystal-lang/crystal/issues/13447

Currently when using `Time#at_beginning_of_week` it assumes the start of the week to be Monday. This works great for many use cases but sometimes we may want to assume the beginning of the week as another day (e.g. Sunday is also commonly considered to be "the beginning of the week")

~I've opted to allow this in 2 forms: With a `Time::DayOfWeek` which is type safe and for convenience, a `String` or `Symbol` that raises an `ArgumentError` if an invalid value is passed~ 
**EDIT:** @Blacksmoke16 pointed out [here](https://github.com/crystal-lang/crystal/pull/13446#issuecomment-1538794685) that we can leverage Symbol autocasting for enum members and avoid the overload altogether

```cr
Time.local.at_beginning_of_week # => 2023-05-08 00:00:00 +00:00
Time.local.at_beginning_of_week(Time::DayOfWeek::Sunday) # => 2023-05-07 00:00:00 +00:00
Time.local.at_beginning_of_week(:sunday) # => 2023-05-07 00:00:00 +00:00
```

This kind of functionality is provided by Rails which is where I got the inspiration from https://api.rubyonrails.org/classes/DateAndTime/Calculations.html#method-i-beginning_of_week

If we decide this is worth having in the standard library I'd also be more than happy to implement it for `Time#at_end_of_week` in this PR or a separate one as well :)

## Docs
### Current
<img width="552" alt="image" src="https://user-images.githubusercontent.com/13454550/236893922-0b196e7b-a862-4307-b401-b3b262f33132.png">
<img width="552" alt="image" src="https://user-images.githubusercontent.com/13454550/236893804-997900db-92d4-47aa-aaa6-b2a08f05ce21.png">

### After this PR
<img width="724" alt="image" src="https://user-images.githubusercontent.com/13454550/236906291-79d7aca8-2c00-48d7-bb10-e275a8e65580.png">
<img width="932" alt="image" src="https://github.com/crystal-lang/crystal/assets/13454550/bcca41d8-8667-417b-9a39-caf16b369926">

Thanks!

Blacksmoke16 commented:
Due to https://crystal-lang.org/reference/1.8/syntax_and_semantics/autocasting.html#symbol-autocasting, I' probably be in favor of dropping support for `String` and just leverage the Eum + Symbol autocasting. Then wouldn't even need the extra overload.

DanielGilchrist commented:
> Due to [crystal-lang.org/reference/1.8/syntax_and_semantics/autocasting.html#symbol-autocasting](https://crystal-lang.org/reference/1.8/syntax_and_semantics/autocasting.html#symbol-autocasting), I'd probably be in favor of dropping support for `String` and just leverage the Enum plus Symbol autocasting. Then wouldn't even need the extra overload.

That's awesome! Thanks for pointing that out. I'll put a commit together now

Sija commented on src/time.cr:
```diff
@@ -1359,9 +1359,20 @@ struct Time
   # TODO: Ensure correctness in local time-line.
-  def at_beginning_of_week : Time
-    (self - (day_of_week.value - 1).days).at_beginning_of_day
+  def at_beginning_of_week(start_day : Time::DayOfWeek = Time::DayOfWeek::Monday) : Time
```
You can use auto-casting already at the definition.

```suggestion
  def at_beginning_of_week(start_day : Time::DayOfWeek = :monday) : Time
```

DanielGilchrist replied:
Thanks!

straight-shoota commented:
Thanks for your contribution, @DanielGilchrist. 🙇
It's very appreciated that you take the initiative. Experience tells that discussions about feature enhancements are better discussed independently of a concrete implementation. Because of this, [the contribution guide](https://github.com/crystal-lang/crystal/blob/master/CONTRIBUTING.md#using-the-issue-tracker) asks to open an issue before sending a pull request.

Now we already have PR and that's great. I'd still ask you to open an issue to propose this change (you can basically take the PR description for that), so we can discuss the feature there.
So long this PR can stay open (ideally in *draft* mode) and we'll get back here when the feature addition is agreed upon.

DanielGilchrist commented:
> Thanks for your contribution, @DanielGilchrist. 🙇 It's very appreciated that you take the initiative. Experience tells that feature enhancements are better discussed independently of a concrete implementation. Because of this, [the contribution guide](https://github.com/crystal-lang/crystal/blob/master/CONTRIBUTING.md?rgh-link-date=2023-05-08T18%3A50%3A27Z#using-the-issue-tracker) asks to open an issue before sending a pull request.
> 
> Now we already have PR and that's great. I'd still ask you to open an issue to propose this change (you can basically take the PR description for that), so we can discuss the feature there. So long this PR can stay open (ideally in _draft_ mode) and we'll get back here when the feature addition is agreed upon.

Thank you for your response and apologies for not reading the contribution guide thoroughly enough 😅 I'll make sure to give it a good read for future contributions

I'll put this PR into draft mode and make an issue now :)

Thanks!

Sija commented on src/time.cr:
````diff
@@ -1359,9 +1359,20 @@ struct Time
   #
+  # The week starts on Monday by default, but can be configured by passing a different `start_day` as a `Time::DayOfWeek`.
+  #
+  # ```
+  # now = Time.local                                  # => 2023-05-08 16:48:13 +00:00
+  # now.at_beginning_of_week(Time::DayOfWeek::Monday) # => 2023-05-08 00:00:00 +00:00
+  # now.at_beginning_of_week(Time::DayOfWeek::Sunday) # => 2023-05-07 00:00:00 +00:00
+  #
+  # # with a symbol
+  # now.at_beginning_of_week(:monday) # => 2023-05-08 00:00:00 +00:00
+  # now.at_beginning_of_week(:sunday) # => 2023-05-07 00:00:00 +00:00
+  # ```
   # TODO: Ensure correctness in local time-line.
````
IMO there's no need to separately showcase the autocasting feature, I'd suggest to simply show it's possible.

```suggestion
  # ```
  # now = Time.local                  # => 2023-05-08 16:48:13 +00:00
  # now.at_beginning_of_week(:monday) # => 2023-05-08 00:00:00 +00:00
  # now.at_beginning_of_week(:sunday) # => 2023-05-07 00:00:00 +00:00
  # ```
  #
  # TODO: Ensure correctness in local time-line.
```

DanielGilchrist replied:
Thank you! I've addressed this here: [`a08f0d2` (#13446)](https://github.com/crystal-lang/crystal/pull/13446/commits/a08f0d22e4832fc5b8dfb395406eac96269007d4)

I also opted to add an example showing the default behaviour without an argument :~)

straight-shoota commented:
https://github.com/crystal-lang/crystal/pull/13446 has had a couple of upvotes (including from two Core Team members), so I think we're good to move on. Sorry for the inconvenience and thanks for following along🙇
This was luckily an uneventful proposal process =)

Simplifying the documentation as suggested in https://github.com/crystal-lang/crystal/pull/13446#discussion_r1187927855 would be good. Then this should be ready to roll.

HertzDevil commented on spec/std/time/time_spec.cr:
There are no specs to show that the hour/minute/second components become 0 after a call to `#at_beginning_of_week`

DanielGilchrist commented:
> #13446 has had a couple of upvotes (including from two Core Team members), so I think we're good to move on. Sorry for the inconvenience and thanks for following along🙇 This was luckily an uneventful proposal process =)

No worries and thank you for your time!

> Simplifying the documentation as suggested in [#13446 (comment)](https://github.com/crystal-lang/crystal/pull/13446#discussion_r1187927855) would be good. Then this should be ready to roll.

I've addressed the remaining comments and relevant specs are passing locally let me know if there are any other changes required before merge :~)

straight-shoota commented on src/time.cr:
````diff
@@ -1359,9 +1359,17 @@ struct Time
+  # The week starts on Monday by default, but can be configured by passing a different `start_day` as a `Time::DayOfWeek`.
+  #
+  # ```
+  # now = Time.local                     # => 2023-05-16 17:53:22 +00:00
+  # now.at_beginning_of_week             # => 2023-05-15 00:00:00 +00:00
+  # now.at_beginning_of_week(:sunday)    # => 2023-05-14 00:00:00 +00:00
+  # now.at_beginning_of_week(:wednesday) # => 2023-05-10 00:00:00 +00:00
````
```suggestion
  # now = Time.utc(2023, 5, 16, 17, 53, 22)
  # now.at_beginning_of_week             # => 2023-05-15 00:00:00 UTC
  # now.at_beginning_of_week(:sunday)    # => 2023-05-14 00:00:00 UTC
  # now.at_beginning_of_week(:wednesday) # => 2023-05-10 00:00:00 UTC
```

straight-shoota reviewed (commented):
I have a small proposal to improve reproduceability of the example by using a fixed timestamp instead of `Time.local`.
